text/0000-remove-shell-processes.md

```diff
@@ -0,0 +1,401 @@
+```
+[[process]]
+type = "hi"
+command = "echo"
+args = ["hello", "world"]
+direct = true
+```
+
+we would convert it to the following in `metadata.toml`
+```
+[[process]]
+type = "hi"
+command = ["echo", "hello", "world"]
+args = []
+api = "0.5"
+direct = true
+```
+This spares users from having to learn about the differences between buildpack APIs in order to predict how additional arguments will behave.
+
+### Evaluating Args
+
+For shell processes provided by older buildpacks, we must continue to Bash evaluate buildpack-provided args to avoid breaking older buildpacks
+
+For example, if a buildpack using API `0.5` creates the following entry in `launch.toml`
+```
+[[process]]
+type = "hi"
+command = "echo"
+args = ["hello", "${WORLD:-world}"]
+```
+
+It should be converted to the following in `metadata.toml`
+
+```
+[[process]]
+type = "hi"
+command = ["echo", "hello", "${WORLD:-world}"]
+args = []
+api = "0.5"
+direct = false
+```
+
+and the launcher should Bash evaluate each entry in `command` to avoid breaking changes.
+
+However, any additional user-provided args should NOT be Bash evaluated to reduce the amount of complexity exposed to end users.
+
+### profiles
+Profiles contributed by older buildpacks will still be evaluated when executing shell process types. But user provided profiles will not be evaluated when using the new platform API, even when a shell process is executed. Again, this is done to prevent differences between the buildpack APIs from leaking into the user interface. Users will only need think about differences between platform APIs.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+In exchange for a reduction in complexity and cognitive overhead buildpack-authors and end users lose certain conveniences like the more intuitive profile interface. This drawback could be remediated by tooling for buildpack authors (to make creation of `exec.d` helpers easy) or by buildpacks (to support `.profile`). However, in the `.profile` case, consistency across ecosystems will be a matter of convention (like Procfile) rather than a guarantee.
+
+# Alternatives
+[alternatives]: #alternatives
+
+## `$(<env>)` syntax
+
+We could use `$(<env>)` for environment variable replacements instead of `${<env>}`. urrently, Kubernetes only replaces `$(FOO)` if `FOO` matches the name of an enivornment variable defined in the `env` section of the PodSpec, otherwise `$(FOO)` is passed through literally. The would allow thelauncher to pick up where kubernetes leaves off and evaluate an remaining references.
+pros:
+* Users can continue to use a syntax they are already familiar with
+cons:
+* User provided variables may be evaluated too soon, before buildpack modifications are applied.
+
+## Lifecycle support for `<app>/.exec`
+When we remove support for `<app>/.profile` we could add support for `<app>/.exec` or similar where `<app>/.exec` must implement the `exec.d` interface.
```

Comment:
I'd like to bring this into scope

Reply:
After thinking long and hard about it I am very opposed to bringing this into scope. I don't think contracts between the app source and the lifecycle are supportable. The platform API should be a contract between a platform and the lifecycle, to enable predictable behavior and a clean upgrade story. Otherwise the same app will behave different on different platforms depending on the platform API in use. I realize the user must have some awareness of the "launch" api to understand argument handling for the resulting image but at least they won't need to change their source code to run on different platforms. This RFC is trying hard to shrink the surface area of platform API, that the app developer needs to be aware of.

If we wanted to respect config like this in the app dir, a buildpack should provide that functionality. App developers can included the exact buildpack they want to get the exact outcome they want without having to keep track of the versions of the platform API currently implemented by the platforms they use (which is entirely outside their control).

Reply:
@ekcasey isn't this already true by us removing support fro `<app>/.profile` since this changes the contract? I get you're trying to simplify/minimize the scope here, but bringing this into scope or not is still changing that contract.

Reply:
While you are right that removing `<app>/.profile` modifies this contract, it shrinks it in a way that should reduce future occurrences of this problem. Adding `<app>/.exec` modified this contract without helping to mitigate this problem in the future.

Reply:
I've created a an RFC that proposes a new mechanism to replace the `.profile` behavior (thus moving the contract to be between a builder<>app). https://github.com/buildpacks/rfcs/pull/175